runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/fn/control/StateRequestHandlerImpl.java

```diff
@@ -0,0 +1,112 @@
+import org.apache.beam.runners.core.StateNamespaces;
+import org.apache.beam.runners.core.StateTags;
+import org.apache.beam.runners.dataflow.worker.ByteStringCoder;
+import org.apache.beam.runners.dataflow.worker.DataflowExecutionContext;
+import org.apache.beam.runners.fnexecution.state.StateRequestHandler;
+import org.apache.beam.sdk.state.BagState;
+import org.apache.beam.sdk.transforms.windowing.GlobalWindow;
+import org.apache.beam.vendor.grpc.v1p13p1.com.google.protobuf.ByteString;
+
+public class StateRequestHandlerImpl implements StateRequestHandler {
+
+  private DataflowExecutionContext.DataflowStepContext ctxt;
+  private ConcurrentHashMap<BeamFnApi.StateKey, BagState<ByteString>> userStateData;
+
+  public StateRequestHandlerImpl(DataflowExecutionContext.DataflowStepContext ctxt) {
+    this.ctxt = ctxt;
+    this.userStateData = new ConcurrentHashMap<>();
+  }
+
+  @Override
+  public CompletionStage<BeamFnApi.StateResponse.Builder> handle(BeamFnApi.StateRequest request)
+      throws Exception {
+    switch (request.getStateKey().getTypeCase()) {
+      case BAG_USER_STATE:
+        return handleBagUserState(request);
+      default:
+        throw new UnsupportedOperationException(
+            String.format(
+                "Dataflow does not handle StateRequests of type %s",
+                request.getStateKey().getTypeCase()));
+    }
+  }
+
+  public void finish() {
+    userStateData.clear();
+  }
+
+  private CompletionStage<BeamFnApi.StateResponse.Builder> handleBagUserState(
+      BeamFnApi.StateRequest request) {
+    BeamFnApi.StateKey.BagUserState bagUserStateKey = request.getStateKey().getBagUserState();
+    // TODO: We should not be required to hold onto a pointer to the bag states for the
+    // user. InMemoryStateInternals assumes that the Java garbage collector does the clean-up work
+    // but instead StateInternals should hold its own references and write out any data and
+    // clear references when the MapTask within Dataflow completes like how WindmillStateInternals
+    // works.
+    BagState<ByteString> state =
+        userStateData.computeIfAbsent(
+            request.getStateKey(),
+            unusedKey ->
+                ctxt.stateInternals()
+                    .state(
+                        // TODO: Once we have access to the ParDoPayload, use its windowing strategy
+                        // to decode the window for the well known window types. Longer term we need
+                        // to swap
+                        // to use the encoded version and not rely on needing to decode the entire
+                        // window.
+                        StateNamespaces.window(GlobalWindow.Coder.INSTANCE, GlobalWindow.INSTANCE),
+                        StateTags.bag(bagUserStateKey.getUserStateId(), ByteStringCoder.of())));
+    switch (request.getRequestCase()) {
+      case GET:
+        return CompletableFuture.completedFuture(
```

Comment:
[optional] better set completable future in switch case, and then utilize single return statement outside. Simplifies readability and understanding of execution flow.

Reply:
Done